LazyGraphModule

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117911

I feel it's easier to open a new PR rather than iterating on the previous PR (https://github.com/pytorch/pytorch/pull/105257 ) since this is more like a rewrite.

In this PR, instead of changing GraphModule directly which can easily causes BC issue, I create a LazyGraphModule class as Zachary & Jason suggested in comments from the previous PR. 

The difference between LazyGraphModule and GraphModule is mainly about how re-compile for the graph module happens. In GraphModule the recompilation happens 'eagerly': constructing a GraphModule will cause the recompilation. While in LazyGraphModule, we just mark the module as needing recompilation. The real recompilation only happens when absolutely required (e.g. call forward method, access the code property etc.). In a lot of cases in torch.compile, the real recompilation eventually is not triggered at all. This can save a few seconds of compilation time.

By default, GraphModule rather than LazyGraphModule is used. `use_lazy_graph_module(True)` context manager can be used to pick LazyGraphModule instead. This has been applied to the torch.compile stack.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

shunting314 committed:
[WIP] LazyGraphModule

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP] LazyGraphModule"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP] LazyGraphModule"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

shunting314 committed:
Update on "LazyGraphModule"


I feel it's easier to open a new PR rather than iterating on the previous PR (https://github.com/pytorch/pytorch/pull/105257 ) since this is more like a rewrite.

In this PR, instead of changing GraphModule directly which can easily causes BC issue, I create a LazyGraphModule class as Zachary & Jason suggested in comments from the previous PR. 

The difference between LazyGraphModule and GraphModule is mainly about how re-compile for the graph module happens. In GraphModule the recompilation happens 'eagerly': constructing a GraphModule will cause the recompilation. While in LazyGraphModule, we just mark the module as needing recompilation. The real recompilation only happens when absolutely required (e.g. call forward method, access the code property etc.). In a lot of cases in torch.compile, the real recompilation eventually is not triggered at all. This can save a few seconds of compilation time.

By default, GraphModule rather than LazyGraphModule is used. `use_lazy_graph_module(True)` context manager can be used to pick LazyGraphModule instead. This has been applied to the torch.compile stack.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

shunting314 committed:
Update on "LazyGraphModule"


I feel it's easier to open a new PR rather than iterating on the previous PR (https://github.com/pytorch/pytorch/pull/105257 ) since this is more like a rewrite.

In this PR, instead of changing GraphModule directly which can easily causes BC issue, I create a LazyGraphModule class as Zachary & Jason suggested in comments from the previous PR. 

The difference between LazyGraphModule and GraphModule is mainly about how re-compile for the graph module happens. In GraphModule the recompilation happens 'eagerly': constructing a GraphModule will cause the recompilation. While in LazyGraphModule, we just mark the module as needing recompilation. The real recompilation only happens when absolutely required (e.g. call forward method, access the code property etc.). In a lot of cases in torch.compile, the real recompilation eventually is not triggered at all. This can save a few seconds of compilation time.

By default, GraphModule rather than LazyGraphModule is used. `use_lazy_graph_module(True)` context manager can be used to pick LazyGraphModule instead. This has been applied to the torch.compile stack.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

shunting314 commented:
[Initial test](https://hud.pytorch.org/benchmark/compilers?startTime=Mon%2C%2015%20Jan%202024%2018%3A29%3A50%20GMT&stopTime=Mon%2C%2022%20Jan%202024%2018%3A29%3A50%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/shunting314/91/head&lCommit=ff96974177ee399431d74d610ed8e6b07f7f44f3&rBranch=main&rCommit=aaae2d8bb6a948dbeab5ac9313453d13d0a09b07) shows a few seconds compilation time reduction (4 seconds for HF). But I also see perf for TB drop from 1.88x -> 1.82x. This should be real since I can repro the drop for basic_gnn_gin. 

By running basic_gnn_gin on devserver, it shows disabling or enabling LazyGraphModule results in exactly the same wrappers generated by inductor. The different perf much be from the python glue code generated by dynamo I guess. I'm looking more into that.

ezyang commented on torch/fx/lazy_graph_module.py:
```diff
@@ -0,0 +1,73 @@
+
+    @classmethod
+    def recompile(cls):
+        cls.forward = cls._lazy_forward
```
This is extremely suspicious. Calling recompile here modifies the CLASS, which will affect ALL FX graphs, not just the instance you are working with.

shunting314 replied:
It does not affect every Fx graph/GraphModule. The reason is because GraphModule has a special `__new__` method https://github.com/pytorch/pytorch/blob/bb28965924cd142a9d12f475fa646c45acb8639f/torch/fx/graph_module.py#L335  which creates a new subclass for every new Fx GraphModule instance.

shunting314 replied:
btw, GraphModule.recompile also update `cls.forward` directly.

shunting314 commented:
> [Initial test](https://hud.pytorch.org/benchmark/compilers?startTime=Mon%2C%2015%20Jan%202024%2018%3A29%3A50%20GMT&stopTime=Mon%2C%2022%20Jan%202024%2018%3A29%3A50%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/shunting314/91/head&lCommit=ff96974177ee399431d74d610ed8e6b07f7f44f3&rBranch=main&rCommit=aaae2d8bb6a948dbeab5ac9313453d13d0a09b07) shows a few seconds compilation time reduction (4 seconds for HF). But I also see perf for TB drop from 1.88x -> 1.82x. This should be real since I can repro the drop for basic_gnn_gin.

> By running basic_gnn_gin on devserver, it shows disabling or enabling LazyGraphModule results in exactly the same wrappers generated by inductor. The different perf much be from the python glue code generated by dynamo I guess. I'm looking more into that.

I think I figured this out. I capture profile with LazyGraphModule enabled and disabled and compare them. When LazyGraphModule is enabled, this part of the profile happen for each iteration of running the compiled model. It is suspicious and does not occur when LazyGraphModule is disabled <img width="671" alt="Screenshot 2024-01-22 at 9 57 53 PM" src="https://github.com/pytorch/pytorch/assets/52589240/3e406780-800c-4ba2-8507-69e941e60fbe">

It looks like we store LazyGraphModule.forward somewhere before the real recompilation happens. So running that stored forward method cause recompilation every time.

By adding debugging code as follows in LazyGraphModule:
```
    def __getattribute__(self, name):
        ret = super().__getattribute__(name)
        if name == "forward":
            if "_lazy_forward" in str(ret):
                breakpoint()
        return ret
```
I found this is caused by returning gm.forward directly in `compile_fx_inner` directly when the graph is empty here: https://github.com/pytorch/pytorch/blob/d84173c025313143c20b8c7c5ffe8b2a24c74216/torch/_inductor/compile_fx.py#L282 . Forcing a recompilation before returning gm.forward fixes the regress. It's a bit surprising that recompiling an empty GraphModule cause noticeable perf difference though.

I'll add more tests to cover this case.

shunting314 committed:
Update on "LazyGraphModule"


I feel it's easier to open a new PR rather than iterating on the previous PR (https://github.com/pytorch/pytorch/pull/105257 ) since this is more like a rewrite.

In this PR, instead of changing GraphModule directly which can easily causes BC issue, I create a LazyGraphModule class as Zachary & Jason suggested in comments from the previous PR. 

The difference between LazyGraphModule and GraphModule is mainly about how re-compile for the graph module happens. In GraphModule the recompilation happens 'eagerly': constructing a GraphModule will cause the recompilation. While in LazyGraphModule, we just mark the module as needing recompilation. The real recompilation only happens when absolutely required (e.g. call forward method, access the code property etc.). In a lot of cases in torch.compile, the real recompilation eventually is not triggered at all. This can save a few seconds of compilation time.

By default, GraphModule rather than LazyGraphModule is used. `use_lazy_graph_module(True)` context manager can be used to pick LazyGraphModule instead. This has been applied to the torch.compile stack.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 commented on torch/fx/graph_module.py:
```diff
@@ -810,7 +809,8 @@ def __deepcopy__(self, memo):
 
     def __copy__(self):
-        res = GraphModule(self, self.graph)
+        from .lazy_graph_module import get_graph_module_cls
+        res = get_graph_module_cls()(self, self.graph)
```
It's weird for a copy to return a different class
```suggestion
        res = self.__class__(self, self.graph)
```

shunting314 replied:
I'll make the change. The reason I do it in the first place is to ensure we get a LazyGraphModule by copying a (non lazy) GraphModule.  Now I've made sure PT2 stack generate LazyGraphModule in the first place, I'll make the change you suggested to make the behavior less confusing.

shunting314 replied:
hmm, it turns out we can not use self.__class__ to construct the copy.

self.__class__ actually is not GraphModule but GraphModuleImpl (defined in GraphModule.__new__). 

Using self.__class__ to construct the copy cause some test failures: 
```
python test/test_fx.py -k test_graph_module_replicate_for_dp
```

complaining we can not access _backend_hooks attribute from the return value of gm.__copy__() .  Looks like the returned GraphModule is not well constructed.

I'll revert this part of change just FYI, @peterbell10

peterbell10 replied:
I guess that's why `__deepcopy__` does this strange dance then:
https://github.com/pytorch/pytorch/blob/dd5e9c64208bf4103abad4866c1f52bbdcd49c88/torch/fx/graph_module.py#L789

presumably that followed by an explicit `__init__` call would work.

shunting314 replied:
I didn't since I feel writing the code this way makes it a bit hard to understand. Also considering the original implementation of `GraphModule.__copy__` already return instance of a different class (since each GraphModule instance has its own class), I feel the current code may not look too bad. 

I can follow up if there is strong reason for the change.

peterbell10 commented on torch/fx/_symbolic_trace.py:
```diff
@@ -693,6 +694,12 @@ def trace(
+                # do real recompilation for LazyGraphModule before retracing
+                from torch.fx.lazy_graph_module import LazyGraphModule
+                if isinstance(root, LazyGraphModule):
+                    root.real_recompile()
```
Why do we sometimes need to manually trigger recompiles?

shunting314 replied:
It's mainly about retracing. Say we have got a LazyGraphModule and we want to retrace that with `symbolic_trace`. Then the `_lazy_forward` method will be traced. `symbolic_trace` does not support that and I got error: https://gist.github.com/shunting314/75549c2e82ae07ac1139c94a3583d259 . I think it's much easier to just force the recompilation before retracing with `symbolic_trace`.

eellison commented:
@shunting314 if it's almost entirely in the pattern matcher that we do all of the recompiles it might be clearer/easier to handle just there and potentially use an explicit invocation.

shunting314 commented:
> @shunting314 if it's almost entirely in the pattern matcher that we do all of the recompiles it might be clearer/easier to handle just there and potentially use an explicit invocation.

So is the suggestion that we do `gm.to_lazy_graph_module()` to convert a graph module to LazyGraphModule at the beginning of pattern matcher?  This may not help much since we already do recompile in `GraphModule.__init__` in this line which can be unneeded:  https://github.com/pytorch/pytorch/blob/58e7ec5843e63ee044e0a4f5aa2583a056a64078/torch/fx/graph_module.py#L428  . But let me know if I misunderstood.

albanD commented on docs/source/conf.py:
```diff
@@ -1060,6 +1060,8 @@
     "reduce_graph_module",
     "reduce_package_graph_module",
+    # torch.fx.lazy_graph_module
+    "use_lazy_graph_module",
```
Adding new entries to this file is not ok.
Please make sure to either document these APIs properly to ensure they show up on the docs website or make them private.

shunting314 replied:
I see. I'll make them private (by prepending an underscore) for now unless people want to add documents instead.

shunting314 commented:
New perf test shows the perf regress on TB has been fixed: [link](https://hud.pytorch.org/benchmark/compilers?startTime=Wed%2C%2017%20Jan%202024%2000%3A06%3A52%20GMT&stopTime=Wed%2C%2024%20Jan%202024%2000%3A06%3A52%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/shunting314/91/head&lCommit=dd5e9c64208bf4103abad4866c1f52bbdcd49c88&rBranch=main&rCommit=bb28965924cd142a9d12f475fa646c45acb8639f).

There are a fews more things I need to do to make CI clean
1. better integration between LazyGraphModule and torchscript. There are more torchscript related failures in CI
2. there is one weird failure: `PYTORCH_TEST_WITH_DYNAMO=1 python test/torch_np/numpy_tests/core/test_numeric.py -k test_boolean` . This does not repro with python 3.10. But if I setup an environment with python 3.11 as the CI job, I can repro the error. Looking more into this.

shunting314 committed:
Update on "LazyGraphModule"


I feel it's easier to open a new PR rather than iterating on the previous PR (https://github.com/pytorch/pytorch/pull/105257 ) since this is more like a rewrite.

In this PR, instead of changing GraphModule directly which can easily causes BC issue, I create a LazyGraphModule class as Zachary & Jason suggested in comments from the previous PR. 

The difference between LazyGraphModule and GraphModule is mainly about how re-compile for the graph module happens. In GraphModule the recompilation happens 'eagerly': constructing a GraphModule will cause the recompilation. While in LazyGraphModule, we just mark the module as needing recompilation. The real recompilation only happens when absolutely required (e.g. call forward method, access the code property etc.). In a lot of cases in torch.compile, the real recompilation eventually is not triggered at all. This can save a few seconds of compilation time.

By default, GraphModule rather than LazyGraphModule is used. `use_lazy_graph_module(True)` context manager can be used to pick LazyGraphModule instead. This has been applied to the torch.compile stack.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 commented:
> So is the suggestion that we do gm.to_lazy_graph_module() to convert a graph module to LazyGraphModule at the beginning of pattern matcher? This may not help much since we already do recompile in GraphModule.__init__ in this line which can be unneeded

We could avoid the `__init__` call if `LazyGraphModule` was implemented as a wrapper around a `GraphModule` rather than inheritance. Would that be feasible? Handle `recompile` specially but forward everything else down to the wrapped module with a `__getattr__` function similar to how ops handlers are written.

Or given that you have to manually call `real_compile` anyway, could you not just have a separate `needs_recompile` variable that gets set by inductor's fx passes? Though that would be harder to enforce that it actually gets used by new passes.

shunting314 commented:
> We could avoid the __init__ call if LazyGraphModule was implemented as a wrapper around a GraphModule rather than inheritance. Would that be feasible? Handle recompile specially but forward everything else down to the wrapped module with a __getattr__ function similar to how ops handlers are written.

I think leveraging composition rather than inheritance still has the problem mentioned earlier. E.g. as AOTAutograd generate fwd/bwd graph for the dynamo generated graph, the GraphModule.recompile method should have already been called once when we construct the graph module for the fwd/bwd graph, but that is NOT needed. Wrap such a GraphModule as an attribute inside a LazyGraphModule does not avoid the unnecessary GraphModule.recompile here. 

Similarly for other graphs we generated for search/replacement patterns. (there are usually much more of them but each of them are smaller)

> Or given that you have to manually call real_compile anyway, could you not just have a separate needs_recompile variable that gets set by inductor's fx passes? 

real_recompile is mainly called in LazyGraphModule._lazy_forward. Other scenarios are quite rare
- inductor return a uncompiled empty GraphModule
- when we serialize a LazyGraphModule (the real use case for this is the Fx graph cache)
- before we retrace a LazyGraphModule with fx symbolic_trace
- I probably need add one more to force the recompilation before torchscripting in order to fix the remaining failures on CI.

shunting314 committed:
Update on "LazyGraphModule"


I feel it's easier to open a new PR rather than iterating on the previous PR (https://github.com/pytorch/pytorch/pull/105257 ) since this is more like a rewrite.

In this PR, instead of changing GraphModule directly which can easily causes BC issue, I create a LazyGraphModule class as Zachary & Jason suggested in comments from the previous PR. 

The difference between LazyGraphModule and GraphModule is mainly about how re-compile for the graph module happens. In GraphModule the recompilation happens 'eagerly': constructing a GraphModule will cause the recompilation. While in LazyGraphModule, we just mark the module as needing recompilation. The real recompilation only happens when absolutely required (e.g. call forward method, access the code property etc.). In a lot of cases in torch.compile, the real recompilation eventually is not triggered at all. This can save a few seconds of compilation time.

By default, GraphModule rather than LazyGraphModule is used. `use_lazy_graph_module(True)` context manager can be used to pick LazyGraphModule instead. This has been applied to the torch.compile stack.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 commented:
Ah okay, I took from your earlier comment that `gm.to_lazy_graph_module()` wouldn't work because it needs to call `GraphModule.__init__` to construct the `LazyGraphModule` instance, which would be solved by using composition. However, if in fact the majority of recompiles occur either outside of the inductor passes, or not on the main graph module then I would agree that composition or having a separate `needs_recompile` flag doesn't help.

shunting314 committed:
Update on "LazyGraphModule"


I feel it's easier to open a new PR rather than iterating on the previous PR (https://github.com/pytorch/pytorch/pull/105257 ) since this is more like a rewrite.

In this PR, instead of changing GraphModule directly which can easily causes BC issue, I create a LazyGraphModule class as Zachary & Jason suggested in comments from the previous PR. 

The difference between LazyGraphModule and GraphModule is mainly about how re-compile for the graph module happens. In GraphModule the recompilation happens 'eagerly': constructing a GraphModule will cause the recompilation. While in LazyGraphModule, we just mark the module as needing recompilation. The real recompilation only happens when absolutely required (e.g. call forward method, access the code property etc.). In a lot of cases in torch.compile, the real recompilation eventually is not triggered at all. This can save a few seconds of compilation time.

By default, GraphModule rather than LazyGraphModule is used. `use_lazy_graph_module(True)` context manager can be used to pick LazyGraphModule instead. This has been applied to the torch.compile stack.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

shunting314 committed:
Update on "LazyGraphModule"


I feel it's easier to open a new PR rather than iterating on the previous PR (https://github.com/pytorch/pytorch/pull/105257 ) since this is more like a rewrite.

In this PR, instead of changing GraphModule directly which can easily causes BC issue, I create a LazyGraphModule class as Zachary & Jason suggested in comments from the previous PR. 

The difference between LazyGraphModule and GraphModule is mainly about how re-compile for the graph module happens. In GraphModule the recompilation happens 'eagerly': constructing a GraphModule will cause the recompilation. While in LazyGraphModule, we just mark the module as needing recompilation. The real recompilation only happens when absolutely required (e.g. call forward method, access the code property etc.). In a lot of cases in torch.compile, the real recompilation eventually is not triggered at all. This can save a few seconds of compilation time.

By default, GraphModule rather than LazyGraphModule is used. `use_lazy_graph_module(True)` context manager can be used to pick LazyGraphModule instead. This has been applied to the torch.compile stack.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

shunting314 commented on torch/_dynamo/eval_frame.py:
```diff
@@ -343,6 +343,14 @@ def get_compiler_config():
+            # optimize a _LazyGraphModule.
+            from torch.fx.lazy_graph_module import _LazyGraphModule
+
+            _LazyGraphModule.force_recompile(fn)
```
This `force_recompile` does not introduce any overhead since we need call forward below anyways which will trigger the real recompilation. 

On the other hand, without this line, we will fail the test
```
PYTORCH_TEST_WITH_DYNAMO=1 python test/torch_np/numpy_tests/core/test_numeric.py -k test_boolean
```
on py3.11. It works on py3.10.

I don't know exactly why py3.11 can not handle it though.

peterbell10 replied:
Should it be the case that accessing `gm.forward` triggers a recompile before you even call it. That would avoid this `force_recompile` and the one in inductor, but I'm not sure if it would introduce unnecessary compiles. wdyt?

peterbell10 commented on torch/_inductor/compile_fx.py:
```diff
@@ -266,6 +273,11 @@ def compile_fx_inner(
+        from torch.fx.lazy_graph_module import _LazyGraphModule
+
+        _LazyGraphModule.force_recompile(gm)
         return make_boxed_func(gm.forward)
```
Should it be the case that accessing `gm.forward` triggers a recompile before you even call it. That would avoid this `force_recompile`, but I'm not sure if it would introduce unnecessary compiles. wdyt?

shunting314 replied:
Hmm, there is one drawback here.

I could add an `__getattribute__` method to `_LazyGraphModule` to do the real recompilation if the attribute name is 'forward'. But in some cases, people retrieve the forward method via the class rather than the object. So `_LazyGraphModule.__getattribute__` will not be called in that case. Here is an example:

https://github.com/pytorch/pytorch/blob/9768f73cb2a0f1ca88739ff904caa5cbd431f3af/torch/fx/_symbolic_trace.py#L702

So now I'm leaning more to keep the old implementation. wdyt?

peterbell10 replied:
Welp, that's unfortunate.

peterbell10 commented on torch/fx/lazy_graph_module.py:
```diff
@@ -0,0 +1,122 @@
+        than 'self.recompile' since for a _LazyGraphModule, self.recompile just
+        mark the need of recompilation and does not return the PythonCode object.
+        """
+        python_code = self._real_recompile()
```
Looks like `__reduce_deploy__` and `__reduce_package__` also use the return value of `recompile`. 

I wonder if there are other users of `recompile()` return value that this code will break. Maybe you could have a `LazyPythonCode` object that runs `_real_compile` when you access its attributes?

shunting314 replied:
> Maybe you could have a LazyPythonCode object that runs _real_compile on the first attribute access?

That's roughly what the GraphModule.code property is doing?

> Looks like __reduce_deploy__ and __reduce_package__ also use the return value of recompile.

Good cache. Although I can not think of a case in PT2 that use torch package/deploy. I'll enhance the code and try to add some tests though.

jansel commented on torch/fx/lazy_graph_module.py:
```diff
@@ -0,0 +1,122 @@
+
+
+@compatibility(is_backward_compatible=False)
+class _LazyGraphModule(GraphModule):
```
Add a docstring explaining what this is any why it is needed.

shunting314 committed:
Update on "LazyGraphModule"


I feel it's easier to open a new PR rather than iterating on the previous PR (https://github.com/pytorch/pytorch/pull/105257 ) since this is more like a rewrite.

In this PR, instead of changing GraphModule directly which can easily causes BC issue, I create a LazyGraphModule class as Zachary & Jason suggested in comments from the previous PR. 

The difference between LazyGraphModule and GraphModule is mainly about how re-compile for the graph module happens. In GraphModule the recompilation happens 'eagerly': constructing a GraphModule will cause the recompilation. While in LazyGraphModule, we just mark the module as needing recompilation. The real recompilation only happens when absolutely required (e.g. call forward method, access the code property etc.). In a lot of cases in torch.compile, the real recompilation eventually is not triggered at all. This can save a few seconds of compilation time.

By default, GraphModule rather than LazyGraphModule is used. `use_lazy_graph_module(True)` context manager can be used to pick LazyGraphModule instead. This has been applied to the torch.compile stack.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

shunting314 committed:
Update on "LazyGraphModule"


I feel it's easier to open a new PR rather than iterating on the previous PR (https://github.com/pytorch/pytorch/pull/105257 ) since this is more like a rewrite.

In this PR, instead of changing GraphModule directly which can easily causes BC issue, I create a LazyGraphModule class as Zachary & Jason suggested in comments from the previous PR. 

The difference between LazyGraphModule and GraphModule is mainly about how re-compile for the graph module happens. In GraphModule the recompilation happens 'eagerly': constructing a GraphModule will cause the recompilation. While in LazyGraphModule, we just mark the module as needing recompilation. The real recompilation only happens when absolutely required (e.g. call forward method, access the code property etc.). In a lot of cases in torch.compile, the real recompilation eventually is not triggered at all. This can save a few seconds of compilation time.

By default, GraphModule rather than LazyGraphModule is used. `use_lazy_graph_module(True)` context manager can be used to pick LazyGraphModule instead. This has been applied to the torch.compile stack.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

shunting314 committed:
Update on "LazyGraphModule"


I feel it's easier to open a new PR rather than iterating on the previous PR (https://github.com/pytorch/pytorch/pull/105257 ) since this is more like a rewrite.

In this PR, instead of changing GraphModule directly which can easily causes BC issue, I create a LazyGraphModule class as Zachary & Jason suggested in comments from the previous PR. 

The difference between LazyGraphModule and GraphModule is mainly about how re-compile for the graph module happens. In GraphModule the recompilation happens 'eagerly': constructing a GraphModule will cause the recompilation. While in LazyGraphModule, we just mark the module as needing recompilation. The real recompilation only happens when absolutely required (e.g. call forward method, access the code property etc.). In a lot of cases in torch.compile, the real recompilation eventually is not triggered at all. This can save a few seconds of compilation time.

By default, GraphModule rather than LazyGraphModule is used. `use_lazy_graph_module(True)` context manager can be used to pick LazyGraphModule instead. This has been applied to the torch.compile stack.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

shunting314 commented:
@pytorchbot merge

shunting314 commented:
The result shows up on the nightly perf test [link](https://hud.pytorch.org/benchmark/compilers?startTime=Mon%2C%2022%20Jan%202024%2018%3A38%3A08%20GMT&stopTime=Mon%2C%2029%20Jan%202024%2018%3A38%3A08%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=main&lCommit=bb6eba189f92790f278f1ee2840f3786da02acaa&rBranch=main&rCommit=f8e14f3b46e68a5271a8c57ce749ad8057d77ddd).
- HF 4 seconds compilation time reduction: 94->90s
- TIMM 2 seconds compilation time reduction: 133->131s
- TB: neutral

I think HF has largest compilation time reduction probably because it exposes more SDPA patterns which results in more calls to GraphModule.recompile. Also transformer's multi-layer nature would results in more matched patterns for each encoder/decoder stack.

fegin commented:
@xmfan, after this PR, `compiled_autograd` + DDP does not work for the benchmarks.

shunting314 commented:
> after this PR, compiled_autograd + DDP does not work for the benchmarks.

@fegin do you have a repro or failed test? I can help fix
